### PR TITLE
DCP aliases

### PR DESCRIPTION
**Pull request by Floessie**
Hi,

This PR adds a little facility to describe camera model aliases for DCP profiles. The new dual-illuminant profiles weigh about 1.1MB which forbids duplication, so a better solution was asked for in #4500.

HTH,
Flössie

### CODE REVIEW

File: rtdata/dcpprofiles/camera_model_aliases.json [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+    "Canon EOS 500D": ["Canon EOS Kiss X3", "Canon EOS Rebel T1i"],
+    "Canon EOS 550D": ["Canon EOS Kiss X4", "Canon EOS Rebel T2i"],
+    "Canon EOS 600D": ["Canon EOS Kiss X5", "Canon EOS Rebel T3i"],
+    "Canon EOS 650D": ["Canon EOS Kiss X6i", "Canon EOS Rebel T4i"],
+    "Canon EOS 700D": ["Canon EOS Kiss X7i", "Canon EOS Rebel T5i"],
+    "Canon EOS 750D": ["Canon EOS Kiss X8i", "Canon EOS Rebel T6i"],
+    "Canon EOS 760D": ["Canon EOS 8000D", "Canon EOS Rebel T6s"],
+    "Canon EOS 800D": ["Canon EOS Kiss X9i", "Canon EOS Rebel T7i"],
+
+    "Minolta Dynax 5D": ["Minolta Maxxum 5D", "Minolta Alpha 5D"],
```
Comment (Beep6581) [inline rtdata/dcpprofiles/camera_model_aliases.json]: If matching on *nix is case-sensitive, the capitalization here would be a problem.

File: rtdata/dcpprofiles/camera_model_aliases.txt [hunk 2]
```diff
@@ -1,3 +0,0 @@
-Canon EOS 600D;Canon EOS Kiss X5;Canon EOS Rebel T3i
-Canon EOS 1300D;Canon EOS Kiss X80;Canon EOS Rebel T6
-MINOLTA DYNAX 7D;MINOLTA MAXXUM 7D;MINOLTA ALPHA 7D;MINOLTA ALPHA SWEET
```
Comment (Beep6581) [inline rtdata/dcpprofiles/camera_model_aliases.txt:3]: I found that "MINOLTA ALPHA SWEET" is an alias for this camera in ExifTool source code.

File: rtdata/dcpprofiles/camera_model_aliases.json (hunk 1 above)
Comment (Floessie) [reply rtdata/dcpprofiles/camera_model_aliases.json]: Internally, the names are matched all-uppercase, thus case-insensitive. If filename matching was okay before, the aliasing won't change it.

File: rtdata/dcpprofiles/camera_model_aliases.txt (hunk 2 above)
Comment (Floessie) [reply rtdata/dcpprofiles/camera_model_aliases.txt:3]: Yep, that got lost. I'll re-add it. Can you guide me to the relevant ExifTool source file?

Comment (Beep6581) [reply rtdata/dcpprofiles/camera_model_aliases.txt:3]: ExifTool 10.92
```
html/models.html-3108-Lumia 950 XL<br>
html/models.html-3109-Nokia_X2</td></tr>
html/models.html-3110-<tr><td colspan=4>&nbsp;<br>
html/models.html-3111-<a name='Minolta' class=th>Minolta / Konica-Minolta</a>
html/models.html-3112-</td></tr><tr valign='top'><td>
html/models.html:3113:ALPHA SWEET Digital<br>
html/models.html-3114-ALPHA-7 Digital<br>
html/models.html-3115-DG-5W<br>
html/models.html-3116-DYNAX 5D<br>
html/models.html-3117-DYNAX 7D<br>
html/models.html-3118-DiMAGE 5<br>
--
lib/Image/ExifTool/Minolta.pm-989-        PrintConv => \%offOn,
lib/Image/ExifTool/Minolta.pm-990-    },
lib/Image/ExifTool/Minolta.pm-991-    0x0114 => [
lib/Image/ExifTool/Minolta.pm-992-        { #8
lib/Image/ExifTool/Minolta.pm-993-            Name => 'MinoltaCameraSettings5D',
lib/Image/ExifTool/Minolta.pm:994:            Condition => '$self->{Model} =~ /^(DYNAX 5D|MAXXUM 5D|ALPHA SWEET)/',
lib/Image/ExifTool/Minolta.pm-995-            SubDirectory => {
lib/Image/ExifTool/Minolta.pm-996-                TagTable => 'Image::ExifTool::Minolta::CameraSettings5D',
lib/Image/ExifTool/Minolta.pm-997-                ByteOrder => 'BigEndian',
lib/Image/ExifTool/Minolta.pm-998-            },
lib/Image/ExifTool/Minolta.pm-999-        },
```

File: rtdata/dcpprofiles/camera_model_aliases.json (hunk 1 above)
Comment (Beep6581) [reply rtdata/dcpprofiles/camera_model_aliases.json]: If the camera is identified by RT as "Foo BAR" and the DCP file is called "Foo bar.dcp", and RT is running in Linux, will it match?

Comment (Floessie) [reply rtdata/dcpprofiles/camera_model_aliases.json]: AFAIK, yes. But I'll check that...

I'll change the Minolta names to uppercase anyway, if that's the official naming.

Comment (Beep6581) [reply rtdata/dcpprofiles/camera_model_aliases.json]: I can't give examples right now, but I'm quite sure I've seen some cameras identified upper-case by RT and differently by Exiv2 or ExifTool. Would be great if this was not a problem.

**Comment by Floessie** [comment]
I just realized, that I'm missing a `cJSON_Minify()` to cope with the comment lines introduced with the second commit. I'll fix that tonight.

File: rtdata/dcpprofiles/camera_model_aliases.json (hunk 1 above)
Comment (Floessie) [reply rtdata/dcpprofiles/camera_model_aliases.json]: I've checked it: Neither the filename nor the alias matching is case sensitive.

**Comment by Beep6581** [comment]
JSON does not support comments, so technically our `camconst.json` is broken (which has been bothering me for some time - every time I edit it, it lights up like a christmas tree). 

```json
    { // Quality C, intermediate ISO samples missing but safely guessed, aperture scaling measures missing
        "make_model": [ "Canon EOS 5DS R", "Canon EOS 5DS" ],
        //"dcraw_matrix": [ 6848,-1661,-221,-3904,10931,3434,-470,1251,6039 ], // DNG_V9.0 A
    }
```

This can be worked-around by including comments as data:
```json
    {
        "//": "Quality C, intermediate ISO samples missing but safely guessed, aperture scaling measures missing",
        "make_model": [ "Canon EOS 5DS R", "Canon EOS 5DS" ],
        "//": "dcraw_matrix: [ 6848,-1661,-221,-3904,10931,3434,-470,1251,6039 ]",
        "//": "DNG_V9.0 A"
    }
```

Would be great if `camera_model_aliases.json` set an example, then `camconst.json` can follow suite later on. I'm happy to help with this if you like, should I make the changes?

**Comment by Floessie** [comment]
Most JSON parsers support comments. We shouldn't drop that, just because some editor highlights this, shouldn't we? :wink:

**Comment by Beep6581** [comment]
> just because some editor highlights this

Sorry, that is misrepresenting the problem.

The JSON standard does not support comments. Just because some parser doesn't choke on them is not a good reason to knowingly keep writing JSON files which don't follow spec and don't validate.

```
json-glib-validate camconst.json
json-glib-validate: file:///home/morgan/programs/code-rawtherapee/rtengine/camconst.json: error parsing file: <data>:6:97: Parse error: unexpected identifier `or', expected value
```

My suggestion was to keep comments, but to do it properly by storing them the only way they can be stored - as data, and have JSON files which validate.

**Comment by Floessie** [comment]
> Sorry, that is misrepresenting the problem.

> The JSON standard does not support comments. Just because some parser doesn't choke on them is not a good reason to knowingly keep writing JSON files which don't follow spec and don't validate.

Yes, technically you are right. However, this is a utility to help us structuring some information. We could as well make up our own, use XML, CSV or whatever. It's not meant as an exchange format here. We could even call the file `*.json5` and would be conforming to that standard.

The headache I have with "comments as data" is the string escaping (really hard to grasp when simply annotating a piece of information) and the parsing/storage overhead we'll get for data that's unwanted.

**Comment by Floessie** [comment]
This is what it would look like:

```json
{
    "//": {
        "//": "This file is part of RawTherapee.",

        "//": "Copyright (C) 2018 Flössie <floessie.mail@gmail.com>",

        "//": "RawTherapee is free software: you can redistribute it and/or modify",
        "//": "it under the terms of the GNU General Public License as published by",
        "//": "the Free Software Foundation, either version 3 of the License, or",
        "//": "(at your option) any later version.",

        "//": "RawTherapee is distributed in the hope that it will be useful,",
        "//": "but WITHOUT ANY WARRANTY; without even the implied warranty of",
        "//": "MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the",
        "//": "GNU General Public License for more details.",

        "//": "You should have received a copy of the GNU General Public License",
        "//": "along with RawTherapee.  If not, see <http://www.gnu.org/licenses/>.",

        "//": "Source: http://rawspeed.org/CameraSupport.html"
    },

    "Canon EOS 1000D": ["Canon EOS Kiss Digital F", "Canon EOS Digital Rebel XS"],
    "Canon EOS 100D": ["Canon EOS Kiss X7", "Canon EOS Rebel SL1"],
    "Canon EOS 1200D": ["Canon EOS Kiss X70", "Canon EOS Rebel T5"],
    "Canon EOS 1300D": ["Canon EOS Kiss X80", "Canon EOS Rebel T6"],
    "Canon EOS 200D": ["Canon EOS Kiss X9", "Canon EOS Rebel SL2"],
    "Canon EOS 300D": ["Canon EOS Kiss Digital", "Canon EOS Digital Rebel"],
    "Canon EOS 350D": ["Canon EOS Kiss Digital N", "Canon EOS Digital Rebel XT"],
    "Canon EOS 400D": ["Canon EOS Kiss Digital X", "Canon EOS Digital Rebel XTi"],
    "Canon EOS 450D": ["Canon EOS Kiss Digital X2", "Canon EOS Kiss X2", "Canon EOS Digital Rebel XSi"],
    "Canon EOS 500D": ["Canon EOS Kiss X3", "Canon EOS Rebel T1i"],
    "Canon EOS 550D": ["Canon EOS Kiss X4", "Canon EOS Rebel T2i"],
    "Canon EOS 600D": ["Canon EOS Kiss X5", "Canon EOS Rebel T3i"],
    "Canon EOS 650D": ["Canon EOS Kiss X6i", "Canon EOS Rebel T4i"],
    "Canon EOS 700D": ["Canon EOS Kiss X7i", "Canon EOS Rebel T5i"],
    "Canon EOS 750D": ["Canon EOS Kiss X8i", "Canon EOS Rebel T6i"],
    "Canon EOS 760D": ["Canon EOS 8000D", "Canon EOS Rebel T6s"],
    "Canon EOS 800D": ["Canon EOS Kiss X9i", "Canon EOS Rebel T7i"],

    "Minolta Dynax 5D": ["Minolta Maxxum 5D", "Minolta Alpha 5D"],
    "Minolta Dynax 7D": ["Minolta Maxxum 7D", "Minolta Alpha 7D"],

    "Panasonic DC-FZ82": ["Panasonic DMC-FZ80", "Panasonic DMC-FZ85"],
    "Panasonic DC-TZ90": ["Panasonic DC-ZS70", "Panasonic DC-FZ91", "Panasonic DC-FZ92", "Panasonic DC-FZ93"],
    "Panasonic DMC-G8": ["Panasonic DMC-G80", "Panasonic DMC-G81", "Panasonic DMC-G85"],
    "Panasonic DMC-LX15": ["Panasonic DMC-LX9", "Panasonic DMC-LX10"],
    "Panasonic DC-TZ100": ["Panasonic DC-ZS100", "Panasonic DC-ZS110", "Panasonic DC-TZ101", "Panasonic DC-TZ110"],
    "Panasonic DMC-TZ71": ["Panasonic DMC-TZ70", "Panasonic DMC-ZS50"],
    "Panasonic DMC-TZ81": ["Panasonic DMC-TZ80", "Panasonic DMC-TZ85", "Panasonic DMC-ZS60"]
}
```

Be aware, that there should be no comment as an array in the top level section, as array value types are interpreted as aliases here.

**Comment by Beep6581** [comment]
Yes, JSON5 could be an alternative if anything supports it, as it's yet just a "[proposed extension](https://spec.json5.org/)"

I don't think we need to include the GPL preamble in this alias-mapping file since it's just data, and not even our data; like a DCP file, but it just happens to be readable in a text editor.

I agree that including metadata mangled with data is not ideal, but to me it seems like a better solution than breaking the format. It would be nice to be able to make use of modern tool capabilities to see correct syntax highlighting, to correctly auto-indent, and to automatically validate structure and find issues. Maybe it's my OCD. I won't push this any further, you know what's best.

**Comment by Floessie** [comment]
@Beep6581 I've removed the comments and added the "Minolta Alpha Sweet" as an alias for the 5D, not 7D. The original Minolta names are mixed-case, so I kept it that way.

Additionally, I added the `cJSON_Minify()` call, so that comments are possible.

Best,
Flössie

**Comment by Beep6581** [comment]
@Floessie fantastic, thank you very much for this!

**Comment by Beep6581** [comment]
I tested using raw files from every model and variant (except the ones listed at the end) and made changes accordingly.

Notes:
* The [Canon EOS 350D](https://raw.pixls.us/getfile.php/758/nice/Canon%20-%20EOS%20350D%20-%20RAW%20(3:2).CR2) actually identifies itself as "Canon EOS 350D DIGITAL".
* Same as above with [400D](https://raw.pixls.us/getfile.php/1764/nice/Canon%20-%20EOS%20400D%20-%20RAW%20(3:2).CR2).
* The [Canon EOS 300D](https://raw.pixls.us/getfile.php/1308/nice/Canon%20-%20EOS%20Digital%20Rebel%20-%20RAW%20(3:2).CRW) / [Canon EOS Digital Rebel](https://raw.pixls.us/getfile.php/1308/nice/Canon%20-%20EOS%20Digital%20Rebel%20-%20RAW%20(3:2).CRW) show no metadata in RT, but once RT adopts Exiv2 their metadata should become visible.

Aliases I could not confirm due to missing samples (generally those released in Japan):
* Canon EOS Kiss X7
* Canon EOS Kiss Digital
* Canon EOS Kiss Digital X
* Canon EOS Kiss Digital X2, Canon EOS Kiss X2
* Canon EOS Kiss X5
* Canon EOS Kiss X6i
* Canon EOS Kiss X7i
* Canon EOS Kiss X8i
* Canon EOS 8000D
* Canon EOS Kiss X9i
* Canon EOS Kiss Digital F
* Canon EOS Kiss X70
* Canon EOS Kiss X80
* Minolta Maxxum 5D, Minolta Alpha 5D, Minolta Alpha Sweet
* Minolta Maxxum 7D, Minolta Alpha 7D
* Panasonic DMC-FZ80, Panasonic DMC-FZ85
* Panasonic DC-ZS70, Panasonic DC-FZ91, Panasonic DC-FZ92, Panasonic DC-FZ93
* Panasonic DMC-G8
* Panasonic DMC-LX9
* Panasonic DC-TZ100, Panasonic DC-ZS100, Panasonic DC-ZS110, Panasonic DC-TZ101, Panasonic DC-TZ110
* Panasonic DMC-ZS50
* Panasonic DMC-TZ85, Panasonic DMC-ZS60

Final file with changes:
```json
{
    "Canon EOS 100D": ["Canon EOS Kiss X7", "Canon EOS REBEL SL1"],
    "Canon EOS 200D": ["Canon EOS Kiss X9", "Canon EOS Rebel SL2"],
    "Canon EOS 300D": ["Canon EOS Kiss Digital", "Canon EOS Digital Rebel"],
    "Canon EOS 350D": ["Canon EOS 350D DIGITAL", "Canon EOS Kiss Digital N", "Canon EOS DIGITAL REBEL XT"],
    "Canon EOS 400D": ["Canon EOS 400D DIGITAL", "Canon EOS Kiss Digital X", "Canon EOS DIGITAL REBEL XTi"],
    "Canon EOS 450D": ["Canon EOS Kiss Digital X2", "Canon EOS Kiss X2", "Canon EOS DIGITAL REBEL XSi"],
    "Canon EOS 500D": ["Canon EOS Kiss X3", "Canon EOS REBEL T1i"],
    "Canon EOS 550D": ["Canon EOS Kiss X4", "Canon EOS REBEL T2i"],
    "Canon EOS 600D": ["Canon EOS Kiss X5", "Canon EOS REBEL T3i"],
    "Canon EOS 650D": ["Canon EOS Kiss X6i", "Canon EOS REBEL T4i"],
    "Canon EOS 700D": ["Canon EOS Kiss X7i", "Canon EOS REBEL T5i"],
    "Canon EOS 750D": ["Canon EOS Kiss X8i", "Canon EOS Rebel T6i"],
    "Canon EOS 760D": ["Canon EOS 8000D", "Canon EOS Rebel T6s"],
    "Canon EOS 800D": ["Canon EOS Kiss X9i", "Canon EOS Rebel T7i"],
    "Canon EOS 1000D": ["Canon EOS Kiss Digital F", "Canon EOS DIGITAL REBEL XS"],
    "Canon EOS 1200D": ["Canon EOS Kiss X70", "Canon EOS REBEL T5"],
    "Canon EOS 1300D": ["Canon EOS Kiss X80", "Canon EOS Rebel T6"],

    "MINOLTA DYNAX 5D": ["Minolta Maxxum 5D", "Minolta Alpha 5D", "Minolta Alpha Sweet"],
    "MINOLTA DYNAX 7D": ["Minolta Maxxum 7D", "Minolta Alpha 7D"],

    "Panasonic DC-FZ82": ["Panasonic DMC-FZ80", "Panasonic DMC-FZ85"],
    "Panasonic DC-TZ90": ["Panasonic DC-ZS70", "Panasonic DC-FZ91", "Panasonic DC-FZ92", "Panasonic DC-FZ93"],
    "Panasonic DMC-G8": ["Panasonic DMC-G80", "Panasonic DMC-G81", "Panasonic DMC-G85"],
    "Panasonic DMC-LX15": ["Panasonic DMC-LX9", "Panasonic DMC-LX10"],
    "Panasonic DC-TZ100": ["Panasonic DC-ZS100", "Panasonic DC-ZS110", "Panasonic DC-TZ101", "Panasonic DC-TZ110"],
    "Panasonic DMC-TZ71": ["Panasonic DMC-TZ70", "Panasonic DMC-ZS50"],
    "Panasonic DMC-TZ81": ["Panasonic DMC-TZ80", "Panasonic DMC-TZ85", "Panasonic DMC-ZS60"]
}
```

The matching works fantastically, imagine what a mess and waste of space we'd have without it!

@Floessie should I commit the changes?

**Comment by Floessie** [comment]
@Beep6581 Thanks for the extensive testing! :tada: Please go ahead and push your changes. :+1:

**Comment by Beep6581** [comment]
@Floessie +1 for merge.

**Comment by Floessie** [comment]
@Beep6581 I see a "no newline at end of file" warning on 68fc436...

**Comment by Beep6581** [comment]
@Floessie yes, it was not intentional, is that a problem? The file validates fine. Feel free to add one if its needed.

**Comment by Floessie** [comment]
@Beep6581 Ha, first edit through the GitHub web editor. Will merge now.

**Pull request merged**